### PR TITLE
Reduce size of fluentd-gcp image using ubuntu-slim

### PR DESCRIPTION
**Pull request by aledbf**
```
gcr.io/google_containers/fluentd-gcp  1.18   ed70ff4ab587  12 weeks ago        411.5 MB
gcr.io/google_containers/fluentd-gcp  1.19   052dc36b53a9  About a minute ago  258.3 MB
```

### CODE REVIEW

**Comment by aledbf** [comment]
ping @bprashanth

**Comment by bprashanth** [comment]
It's *still* humongous though :(
Maybe Alex has some suggestions about getting rid of more stuff. That apt-get update is a killer.

**Comment by aledbf** [comment]
@bprashanth `293.5MB` now. The directory `/opt/google-fluentd` is `216MB`

**Comment by aledbf** [comment]
Is possible to gain another `22MB` if is possible to remove `/opt/google-fluentd/embedded/share/doc/`

**Comment by a-robinson** [comment]
Thanks @aledbf! You'll also want to update cluster/saltbase/salt/fluent-gcp/fluent-gcp as well so that the image actually gets used.

Getting rid of `/opt/google-fluentd/embedded/share/doc/` sounds good to me. There's a lot of other random junk in the image too (e.g. why is postgres included with fluentd?), but that's a nice easy win.

Have you tested it it to make sure it still works?

Also cc @igorpeshansky for general interest in the fluentd package.

**Comment by a-robinson** [comment]
Also, [the fluentd-es image](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/fluentd-elasticsearch/fluentd-es-image/Dockerfile) has the same problem if you'd like to shrink that down in a similar way, but no worries if you just want to do the fluentd-gcp image.

**Comment by a-robinson** [comment]
@k8s-bot ok to test

**Comment by aledbf** [comment]
Now is `258MB`

**Comment by aledbf** [comment]
@a-robinson is possible to change how the fluentd-es image is build? 
```
gcr.io/google_containers/fluentd-elasticsearch  1.16  ccbce8f9742e        15 minutes ago      480.8 MB
gcr.io/google_containers/fluentd-elasticsearch  1.15  4a18094787db        3 months ago        562 MB
```

Using a build script is possible to remove `make` and `g++` after the fluentd installation 
```
gcr.io/google_containers/fluentd-elasticsearch  1.16  6694b8bf2a31        3 minutes ago       310 MB
```

File: cluster/addons/fluentd-gcp/fluentd-gcp-image/Dockerfile [hunk 1]
```diff
@@ -20,24 +20,35 @@
...
+  /opt/google-fluentd/embedded/lib/postgresql \
+  /opt/google-fluentd/embedded/bin/postgres \
+  /opt/google-fluentd/embedded/share/postgresql \
+  /var/log/google-fluentd
 
 # Install the record reformer plugin.
-RUN /usr/sbin/google-fluentd-gem install fluent-plugin-record-reformer
-
 # Remove the misleading log file that gets generated when the agent is installed
-RUN rm -rf /var/log/google-fluentd
```
Comment (a-robinson) [inline cluster/addons/fluentd-gcp/fluentd-gcp-image/Dockerfile]: Any particular reason you got rid of this? This file caused some confusion amongst people looking into the container's filesystem before we added this step to the build.

**Comment by a-robinson** [comment]
@aledbf - I'm alright with it. cc @jimmidyson in case he's concerned about it.

And since I think it got lost in all my comments up above, have you verified in any way that the new image works?

File: cluster/addons/fluentd-gcp/fluentd-gcp-image/Dockerfile (hunk 1 above)
Comment (aledbf) [inline cluster/addons/fluentd-gcp/fluentd-gcp-image/Dockerfile]: Is being removed here https://github.com/kubernetes/kubernetes/pull/26652/files#diff-77fb165d87e0464cc4f2358f49feebadR47

Comment (a-robinson) [inline cluster/addons/fluentd-gcp/fluentd-gcp-image/Dockerfile]: Thanks. You might want to remove the comment on line 50 then.

Comment (aledbf) [inline cluster/addons/fluentd-gcp/fluentd-gcp-image/Dockerfile]: done

**Comment by aledbf** [comment]
@a-robinson is failing because the image is not available in gcr
`"Tag 1.16 not found in repository gcr.io/google_containers/fluentd-elasticsearch"`

**Comment by bprashanth** [comment]
Actually ubuntu 16 isn't part of the supported distros for the logging agent: https://cloud.google.com/logging/docs/agent/. Maybe we can do something similar with one of the supported images?

**Comment by aledbf** [comment]
@bprashanth done
```
gcr.io/google_containers/fluentd-elasticsearch 1.16  25481f3b983d   2 minutes ago   386.2 MB
gcr.io/google_containers/fluentd-gcp           1.19  38f55dafa2c2   8 minutes ago   368.7 MB
```

**Comment by aledbf** [comment]
ok, now using `ubuntu-slim:0.1` (based on ubuntu:15.10)
```
gcr.io/google_containers/fluentd-elasticsearch  1.16   ac4b9cb8febf  10 seconds ago  314.2 MB
gcr.io/google_containers/fluentd-gcp            1.19   52d970a4b138  3 minutes ago   344.4 MB
```

File: cluster/addons/fluentd-gcp/fluentd-gcp-image/Dockerfile (hunk 1 above)
Comment (a-robinson) [inline cluster/addons/fluentd-gcp/fluentd-gcp-image/Dockerfile]: Looks like it's still there?

"# Remove the misleading log file that gets generated when the agent is installed"

**Comment by a-robinson** [comment]
> @a-robinson is failing because the image is not available in gcr
> "Tag 1.16 not found in repository gcr.io/google_containers/fluentd-elasticsearch"

I've built and pushed new versions of fluentd-elasticsearch and fluentd-gcp for you based on your latest commits.

**Comment by a-robinson** [comment]
@k8s-bot e2e test this please: #IGNORE

**Comment by a-robinson** [comment]
@k8s-bot e2e test this issue: #IGNORE

**Comment by jimmidyson** [comment]
I've no concerns with switching to base image. Quite a modest saving though - switching to Alpine would probably bring greater savings, but not a supported base image from the docs I don't think. But :+1:

**Comment by aledbf** [comment]
@a-robinson I cannot reproduce the error I see in the logs:
`https://10.0.0.1:443/api: SSL_connect returned=1 errno=0 state=error: certificate verify failed"`

**Comment by bprashanth** [comment]
I think that's just complaining that the certificate returned doesn't match any known CA in the trusted root store in the container, or wherever the client is looking for roots

**Comment by a-robinson** [comment]
@aledbf, that would seem to be an issue with the metadata filter plugin not liking the apiserver's certificate. @jimmidyson, know why that would happen?

**Comment by jimmidyson** [comment]
@a-robinson @aledbf There should be no problem as long as the certiifcate is valid for the API server endpoint.

**Comment by a-robinson** [comment]
Does it expect certs that are valid as per normal certificate authorities? Because GCE/GKE deployments of k8s create their own root CA that signs the master's cert.

**Comment by jimmidyson** [comment]
@a-robinson It should use the CA cert distributed to pods at `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt`. I assume that is done on GKE?

**Comment by a-robinson** [comment]
Yes, absolutely. Also, this failing test is just on GCE directly, not GKE.

**Comment by jimmidyson** [comment]
Is this problem happening on any env that I can connect to to investigate or a way for me to easily recreate so I can take a look? I'd probably be checking that certificate is valid for `kubernetes.default.svc` using the CA certificate at `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt`. Should be able to use openssl to verify.

**Comment by a-robinson** [comment]
You should be able to recreate by either patching in @aledbf's changes or just locally changing your setup to use gcr.io/google_containers/fluentd-elasticsearch:1.16 instead of 1.15.

I don't believe anything specific to the GCE setup should be having an effect, so you shouldn't even have to run it on GCE, just any setup that uses fluentd-es.

**Comment by aledbf** [comment]
@a-robinson @jimmidyson I just started a cluster running `go run hack/e2e.go -v --build --up`.
I see the same error that appears in the logs. The strange thing is that if the pod is started manually it works just fine https://gist.github.com/aledbf/91ee2f3e88d9b8f93167a39f016be568
The different from kubectl describe is that in the failing pods there's no `service-account-token`

**Comment by a-robinson** [comment]
@aledbf, I'm sorry this got lost!

@piosz @fgrzadkowski - you might want to look into resurrecting this to shrink the fluentd image size down

**Comment by aledbf** [comment]
@a-robinson np. I just created a new branch to update this PR and I see the same issue, no service account which leads to errors in the ssl certificate and missing bearer token.
https://github.com/kubernetes/kubernetes/compare/master...aledbf:fluent-image-size?expand=1

**Comment by a-robinson** [comment]
I believe that's due to the inclusion of the kubernetes_metadata filter in [the fluentd-elasticsearch config](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/fluentd-elasticsearch/fluentd-es-image/td-agent.conf). Try removing the

```
<filter kubernetes.**>
  type kubernetes_metadata
</filter>
```

stanza from it and seeing what happens.

**Comment by aledbf** [comment]
@a-robinson with those commented the pod works without issues
```
old-mbp:kubernetes aledbf$ kubectl get pods --all-namespaces
NAMESPACE     NAME                                                      READY     STATUS    RESTARTS   AGE
kube-system   elasticsearch-logging-v1-gr72w                            1/1       Running   0          5m
kube-system   elasticsearch-logging-v1-qdxug                            1/1       Running   0          5m
kube-system   etcd-server-e2e-test-aledbf-master                        1/1       Running   0          4m
kube-system   etcd-server-events-e2e-test-aledbf-master                 1/1       Running   0          4m
kube-system   fluentd-elasticsearch-e2e-test-aledbf-minion-group-g5x6   1/1       Running   0          4m
kube-system   fluentd-elasticsearch-e2e-test-aledbf-minion-group-s5b0   1/1       Running   0          4m
kube-system   fluentd-elasticsearch-e2e-test-aledbf-minion-group-w6tm   1/1       Running   0          4m
kube-system   heapster-v1.1.0-527143062-ppgkq                           4/4       Running   0          4m
kube-system   kibana-logging-v1-34pzx                                   1/1       Running   1          5m
kube-system   kube-addon-manager-e2e-test-aledbf-master                 1/1       Running   0          5m
kube-system   kube-apiserver-e2e-test-aledbf-master                     1/1       Running   2          4m
kube-system   kube-controller-manager-e2e-test-aledbf-master            1/1       Running   0          4m
kube-system   kube-dns-v19-5md4y                                        3/3       Running   0          5m
kube-system   kube-proxy-e2e-test-aledbf-minion-group-g5x6              1/1       Running   0          4m
kube-system   kube-proxy-e2e-test-aledbf-minion-group-s5b0              1/1       Running   0          4m
kube-system   kube-proxy-e2e-test-aledbf-minion-group-w6tm              1/1       Running   0          4m
kube-system   kube-scheduler-e2e-test-aledbf-master                     1/1       Running   0          5m
kube-system   kubernetes-dashboard-v1.1.0-wl6lu                         1/1       Running   0          5m
kube-system   l7-default-backend-v1.0-6tmsa                             1/1       Running   0          5m
kube-system   l7-lb-controller-v0.7.0-e2e-test-aledbf-master            1/1       Running   0          5m
kube-system   monitoring-influxdb-grafana-v3-0                          2/2       Running   0          5m
kube-system   node-problem-detector-v0.1-bhxx1                          1/1       Running   0          5m
kube-system   node-problem-detector-v0.1-l46dk                          1/1       Running   0          5m
kube-system   node-problem-detector-v0.1-uudi8                          1/1       Running   0          5m
kube-system   node-problem-detector-v0.1-wzptz                          1/1       Running   0          5m
old-mbp:kubernetes aledbf$ kubectl logs -f fluentd-elasticsearch-e2e-test-aledbf-minion-group-g5x6 --namespace kube-system
2016-08-01 22:17:04 +0000 [info]: reading config file path="/etc/td-agent/td-agent.conf"
2016-08-01 22:17:05 +0000 [info]: starting fluentd-0.12.26
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-mixin-config-placeholders' version '0.4.0'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-mixin-plaintextformatter' version '0.2.6'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-docker_metadata_filter' version '0.1.3'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-elasticsearch' version '1.5.0'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-kubernetes_metadata_filter' version '0.24.0'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-mongo' version '0.7.13'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-rewrite-tag-filter' version '1.5.5'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-s3' version '0.6.8'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-scribe' version '0.10.14'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-td' version '0.10.28'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-td-monitoring' version '0.2.2'
2016-08-01 22:17:05 +0000 [info]: gem 'fluent-plugin-webhdfs' version '0.4.2'
2016-08-01 22:17:05 +0000 [info]: gem 'fluentd' version '0.12.26'
2016-08-01 22:17:05 +0000 [info]: adding match pattern="fluent.**" type="null"
2016-08-01 22:17:05 +0000 [info]: adding match pattern="**" type="elasticsearch"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
2016-08-01 22:17:06 +0000 [info]: adding source type="tail"
....
```

**Pull request closed without merging**